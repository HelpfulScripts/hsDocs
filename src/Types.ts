/**
 * # Types
 */
/** */

import { log as _log } from 'hsutil';import { userInfo } from 'os';
 const log = _log('DocTypes');

export const Counts = <{[docType:string]: {[docSubtype:string]: number}}>{};

interface DocID {
    id:     number;
}

interface DocNameType {
    name:   string;
    type:   string;
}

interface DocTypeNameID extends DocNameType, DocID {
}

interface DocKindNameID extends DocID {
    name:   string;
    kind:               string;
}

export interface DocBaseNode extends DocKindNameID {
    kindString?:        string;
	flags:              DocNodeFlags;
}

interface DocElementType extends DocKindNameID {
    elementType?:   DocNameType;
}

interface DocSignature extends DocBaseNode {
    type:               DocType;
    comment?:           DocComment;
    parameters?:        DocNode[];
    inheritedFrom?:     DocInheritedFrom;
    overwrites?:        DocTypeNameID;
    implementationOf?:  DocImplementationOf;
    typeParameter?:     DocBaseNode[];
}

interface DocNode extends DocSignature {
    originalName?:      string;
    defaultValue?:      string;
    implementedTypes?:  DocImplementedTypes[];
    implementedBy?:     DocImplementedBy[];
    extendedTypes?:     DocExtendedTypes[];
    extendedBy?:        DocExtendedBy[];
    children:           DocNode[];
    groups:             DocReference[];
    sources?:           DocSource[];
    signatures?:        DocSignature[];
    indexSignature?:    DocNode[];
    getSignature?:      DocNode[];
}

interface DocType extends DocTypeNameID {
    declaration:    DocNode;
    elementType?:   DocType;
    elements?:      DocElement[];
    typeArguments?: DocType[];
    types?:         DocType[];
    value?:         string;
}

interface DocNodeFlags {
    isExported?:            boolean;
    isPrivate?:             boolean;
    isPublic?:              boolean;
    isStatic?:              boolean;
    isConst?:               boolean;
    isLet?:                 boolean;
    isOptional?:            boolean;
    isProtected?:           boolean;
    isAbstract?:            boolean;
    isRest?:                boolean;
    isConstructorProperty?: boolean;
}

interface DocSource {
    fileName:   string;
	line:       number;
	character:  number;
}

interface DocComment {
    shortText?: string;
    text?:      string;
    tags?:      DocTag[];
    returns?:   string;

}

interface DocTypeArguments extends DocNameType {
    elementType: DocNameType;
}

interface DocElementType extends DocTypeNameID {
    declaration:    DocNode;
}

interface DocElement extends DocTypeNameID {
}

interface DocReference {
    title:      string;
	kind:       number;
	children:   number[];
}

interface DocTag {
    tag:        string;
	text:       string;
    param:      string;
}

interface DocInheritedFrom extends DocTypeNameID {
}

interface DocExtendedTypes extends DocTypeNameID {
}

interface DocExtendedBy extends DocTypeNameID {
}

interface DocImplementationOf extends DocTypeNameID {
}

interface DocImplementedTypes extends DocTypeNameID {
}

interface DocImplementedBy extends DocTypeNameID {
}


//-----------------

const accumulate = (acc:number, curr:number) => acc+curr;

const use = (docType:string, docSubtype:string) => { 
    Counts[docType] = Counts[docType] || {};
    Counts[docType][docSubtype] = (Counts[docType][docSubtype] || 0) + 1;
};


export function testNode(node:DocNode, name=''):typeof Counts {
    testFullNode(node);
    return Counts;
}

export function testFullNode(node:DocNode, name=''):number {
    if (name==='') { name = node.name; }
    const num =  Object.getOwnPropertyNames(node).map((k):number => {
        use('DocNode', k);
        switch (k) {
            case 'lib':             return 0;  // added by HsDocs
            case 'modules':         return 0;  // added by HsDocs
            case 'fullPath':        return 0;  // added by HsDocs
            case 'id':              return testNumber(k, node.id, node, node.name);
            case 'name':            return testString(k, node.name, node,`${name}-${k}`);
            case 'kind':            return testNumber(k, node.kind, node,`${name}-${k}`);
            case 'kindString':      return testString(k, node.kindString, node,`${name}-${k}`);
            case 'flags':           return testFlags(node.flags, node,`${name}-${k}`);
            case 'originalName':    return testString(k, node.originalName, node,`${name}-${k}`);
            case 'defaultValue':    return testString(k, node.defaultValue, node,`${name}-${k}`);
            case 'comment':         return testComment(node.comment, node,`${name}-${k}`);
            case 'sources':         return testSources(node.sources, node,`${name}-${k}`);
            case 'type':            return testType(node.type, node,`${name}-${k}`);
            case 'inheritedFrom':   return testTypeNameID(node.inheritedFrom, node,`${name}-${k}`);
            case 'implementationOf':return testTypeNameID(node.implementationOf, node,`${name}-${k}`);
            case 'overwrites':      return testTypeNameID(node.overwrites, node,`${name}-${k}`);
            case 'implementedTypes':return node.implementedTypes.map(t => testTypeNameID(t, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'implementedBy':   return node.implementedBy.map(t => testTypeNameID(t, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'extendedBy':      return node.extendedBy.map(e => testTypeNameID(e, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'extendedTypes':   return node.extendedTypes.map(t =>testTypeNameID(t, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'children':        return node.children.map(c => testFullNode(c, `${name}-${c.name}`)).reduce(accumulate, 0);
            case 'groups':          return node.groups.map(c => testReference(c, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'signatures':      return node.signatures.map(c => testSignature(c, node, `${name}-${k}`)).reduce(accumulate, 0);
            case 'parameters':      return node.parameters.map(c => testFullNode(c, `${name}-${k}`)).reduce(accumulate, 0);
            case 'indexSignature':  return node.indexSignature.map(is => testFullNode(is, `${name}-${k}`)).reduce(accumulate, 0);
            case 'getSignature':    return node.getSignature.map(is => testFullNode(is, `${name}-${k}`)).reduce(accumulate, 0);
            case 'typeParameter':   return node.typeParameter.map(is => testBaseNode(is, node, `${name}-${k}`)).reduce(accumulate, 0);
            default: log.warn(`testFullNode found unknown key '${k}' for path '${name}' ${node['lib']?'in lib '+node['lib'] : ''}`);
        }
        return -1;
    }).reduce(accumulate, 0);
    log.debug(`${num} tests in node ${node.name}`);
    return num;
}

function testSignature(test:DocSignature, node: DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocSignature', k);
        switch (k) {
            case 'lib':             return 0;  // added by HsDocs
            case 'fullPath':        return 0;  // added by HsDocs
            case 'id':              return testNumber(k, test.id, node, node.name);
            case 'name':            return testString(k, test.name, node,`${name}-${k}`);
            case 'kind':            return testNumber(k, test.kind, node,`${name}-${k}`);
            case 'kindString':      return testString(k, test.kindString, node,`${name}-${k}`);
            case 'flags':           return testFlags(test.flags, node,`${name}-${k}`);
            case 'comment':         return testComment(test.comment, node,`${name}-${k}`);
            case 'parameters':      return test.parameters.map(c => testFullNode(c, `${name}-${k}`)).reduce(accumulate, 0);
            case 'type':            return testType(test.type, node,`${name}-${k}`);
            case 'inheritedFrom':   return testTypeNameID(test.inheritedFrom, node,`${name}-${k}`);
            case 'overwrites':      return testTypeNameID(test.overwrites, node,`${name}-${k}`);
            case 'implementationOf':return testTypeNameID(test.implementationOf, node,`${name}-${k}`);
            case 'typeParameter':   return test.typeParameter.map(is => testBaseNode(is, node, `${name}-${k}`)).reduce(accumulate, 0);
            default: log.warn(`testNameType found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

function testBaseNode(test:DocBaseNode, node: DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocBaseNode', k);
        switch (k) {
            case 'id':              return testNumber(k, test.id, node, node.name);
            case 'name':            return testString(k, test.name, node,`${name}-${k}`);
            case 'kind':            return testNumber(k, test.kind, node,`${name}-${k}`);
            case 'kindString':      return testString(k, test.kindString, node,`${name}-${k}`);
            case 'flags':           return testFlags(test.flags, node,`${name}-${k}`);
            default: log.warn(`testBaseNode found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}



function testNameType(test:DocNameType, node: DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocNameType', k);
        switch (k) {
            case 'type':    return testString(k, test.type, node,`${name}-${k}`);
            case 'name':    return testString(k, test.name, node,`${name}-${k}`);
            default: log.warn(`testNameType found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

function testTypeNameID(test:DocTypeNameID, node: DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocTypeNameID', k);
        switch (k) {
            case 'type':    return testString(k, test.type, node,`${name}-${k}`);
            case 'name':    return testString(k, test.name, node,`${name}-${k}`);
            case 'id':      return testNumber(k, test.id, node,`${name}-${k}`);
            default: log.warn(`testTypeNameID found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

function testFlags(test:DocNodeFlags, node:DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocNodeFlags', k);
        switch (k) {
            case 'isExported':  return testBoolean(k, test.isExported, node,`${name}-${k}`);
            case 'isPublic':    return testBoolean(k, test.isPublic, node,`${name}-${k}`);
            case 'isPrivate':   return testBoolean(k, test.isPrivate, node,`${name}-${k}`);
            case 'isStatic':    return testBoolean(k, test.isStatic, node,`${name}-${k}`);
            case 'isConst':     return testBoolean(k, test.isConst, node,`${name}-${k}`);
            case 'isLet':       return testBoolean(k, test.isLet, node,`${name}-${k}`);
            case 'isOptional':  return testBoolean(k, test.isOptional, node,`${name}-${k}`);
            case 'isProtected': return testBoolean(k, test.isProtected, node,`${name}-${k}`);
            case 'isAbstract':  return testBoolean(k, test.isAbstract, node,`${name}-${k}`);
            case 'isRest':      return testBoolean(k, test.isRest, node,`${name}-${k}`);
            case 'isConstructorProperty': return testBoolean(k, test.isConstructorProperty, node,`${name}-${k}`);
            default: log.warn(`testFlags found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

function testComment(test:DocComment, node:DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocComment', k);
        switch (k) {
            case 'shortText':   return testString(k, test.shortText, node,`${name}-${k}`);
            case 'text':        return testString(k, test.text, node,`${name}-${k}`);
            case 'tags':        return test.tags.map(t => testTag(t, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'returns':     return testString(k, test.returns, node,`${name}-${k}`);
            default: log.warn(`testComment found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

function testSources(test:DocSource[], node: DocNode, name: string):number {
    return test.map(s => Object.getOwnPropertyNames(s).map(k => {
        use('DocSource', k);
        switch (k) {
            case 'fileName':    return testString(k, s.fileName, node,`${name}-${k}`);
            case 'line':        return testNumber(k, s.line, node,`${name}-${k}`);
            case 'character':   return testNumber(k, s.character, node,`${name}-${k}`);
            default: log.warn(`testSources found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0)).reduce(accumulate, 0);
}

function testType(test:DocType, node: DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocType', k);
        switch (k) {
            case 'type':            return testString(k, test.type, node,`${name}-${k}`);
            case 'name':            return testString(k, test.name, node,`${name}-${k}`);
            case 'id':              return testNumber(k, test.id, node,`${name}-${k}`);
            case 'value':           return testString(k, test.value, node,`${name}-${k}`);
            case 'declaration':     return testFullNode(test.declaration,`${name}-${k}`);
            case 'elementType':     return testType(test.elementType, node,`${name}-${k}`);
            case 'elements':        return test.elements.map(t => testTypeNameID(t, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'typeArguments':   return test.typeArguments.map(t => testType(t, node,`${name}-${k}`)).reduce(accumulate, 0);
            case 'types':           return test.types.map((t:DocType) => testType(t, node,`${name}-${k}`)).reduce(accumulate, 0);
            default: log.warn(`testType found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

function testReference(test:DocReference, node:DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocReference', k);
        switch (k) {
            case 'title':       return testString(k, test.title, node,`${name}-${k}`);
            case 'kind':        return testNumber(k, test.kind, node,`${name}-${k}`);
            case 'children':    return test.children.map(c => testNumber(k, c, node,`${name}-${k}`)).reduce(accumulate, 0);
            default: log.warn(`testReference found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

function testTag(test:DocTag, node:DocNode, name: string):number {
    return Object.getOwnPropertyNames(test).map(k => {
        use('DocTag', k);
        switch (k) {
            case 'tag':     return testString(k, test.tag, node,`${name}-${k}`);
            case 'text':    return testString(k, test.text, node,`${name}-${k}`);
            case 'param':   return testString(k, test.param, node,`${name}-${k}`);
            default: log.warn(`testTag found unknown key '${k}' for path '${name}'`);
        }
        return -1;
    }).reduce(accumulate, 0);
}

// function testArguments(test:DocTypeArguments, node:DocNode, name: string):number {
//     return Object.getOwnPropertyNames(test).map(k => {
//         use('DocTypeArguments', k);
//         switch (k) {
//             case 'type':        return testString(k, test.type, node,`${name}-${k}`);
//             case 'name':        return testString(k, test.name, node,`${name}-${k}`);
//             case 'elementType': return testNameType(test.elementType, node,`${name}-${k}`);
//             default: log.warn(`testArguments found unknown key '${k}' for path '${name}'`);
//         }
//         return -1;
//     }).reduce(accumulate, 0);
// }

// function testElementType(test:DocElementType, node:DocNode, name: string):number {
//     return Object.getOwnPropertyNames(test).map(k => {
//         use('DocElementType', k);
//         switch (k) {
//             case 'id':          return testNumber(k, test.id, node,`${name}-${k}`);
//             case 'name':        return testString(k, test.name, node,`${name}-${k}`);
//             case 'type':        return testString(k, test.type, node,`${name}-${k}`);
//             case 'declaration': return testFullNode(test.declaration,`${name}-${k}`);
//             default: log.warn(`testElementType found unknown key '${k}' for path '${name}'`);
//         }
//         return -1;
//     }).reduce(accumulate, 0);
// }

//----------


function testString(k:string, v:any, node:DocNode, name: string):number {
    if (typeof v !== 'string') { log.warn(`value ${v} for key '${k}' is not a string for path '${name}'`); }
    else { log.debug(`test ${k}:${v.slice(0, 10)} as string for path '${name}'`); }
    return 1;
}

function testBoolean(k:string, v:any, node:DocNode, name: string):number {
    if (typeof v !== 'boolean') { log.warn(`value ${v} for key '${k}' is not a boolean for path '${name}'`); }
    else { log.debug(`test ${k}:${v} as string for path '${name}'`); }
    return 1;
}

function testNumber(k:string, v:any, node:DocNode, name: string):number {
    if (typeof v !== 'number') { log.warn(`value ${v} for key '${k}' is not a number for path '${name}'`); }
    else { log.debug(`test ${k}:${v} as string for path '${name}'`); }
    return 1;
}
