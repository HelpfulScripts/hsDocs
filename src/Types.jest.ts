

import { node }     from 'hsnode';
import { testNode } from './Types';



describe('Types', () => {
    test('hsDocs', async () => {
        const content = await node.fs.readJsonFile('./src/test/hsDocs.json');
        expect(testNode(content)).toMatchSnapshot();
    });
    test('hsGraphD3', async () => {
        const content = await node.fs.readJsonFile('./src/test/hsGraphD3.json');
        expect(testNode(content)).toMatchSnapshot();
    });
    test('hsLayout', async () => {
        const content = await node.fs.readJsonFile('./src/test/hsLayout.json');
        expect(testNode(content)).toMatchSnapshot();
    });
    test('hsWidget', async () => {
        const content = await node.fs.readJsonFile('./src/test/hsWidget.json');
        expect(testNode(content)).toMatchSnapshot();
    });
});
