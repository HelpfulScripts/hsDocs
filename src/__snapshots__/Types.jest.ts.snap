// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Types hsDocs 1`] = `
Object {
  "DocComment": Object {
    "returns": 7,
    "shortText": 47,
    "text": 44,
  },
  "DocNode": Object {
    "children": 64,
    "comment": 59,
    "defaultValue": 45,
    "extendedBy": 6,
    "extendedTypes": 18,
    "flags": 503,
    "groups": 64,
    "id": 503,
    "indexSignature": 3,
    "inheritedFrom": 63,
    "kind": 503,
    "kindString": 502,
    "name": 503,
    "originalName": 13,
    "overwrites": 5,
    "parameters": 3,
    "signatures": 86,
    "sources": 332,
    "type": 364,
  },
  "DocNodeFlags": Object {
    "isConst": 21,
    "isExported": 74,
    "isOptional": 55,
    "isPrivate": 3,
    "isProtected": 3,
    "isPublic": 3,
    "isStatic": 5,
  },
  "DocReference": Object {
    "children": 86,
    "kind": 86,
    "title": 86,
  },
  "DocSignature": Object {
    "comment": 30,
    "flags": 86,
    "id": 86,
    "inheritedFrom": 6,
    "kind": 86,
    "kindString": 86,
    "name": 86,
    "overwrites": 3,
    "parameters": 82,
    "type": 86,
  },
  "DocSource": Object {
    "character": 333,
    "fileName": 333,
    "line": 333,
  },
  "DocType": Object {
    "declaration": 10,
    "elementType": 38,
    "id": 81,
    "name": 445,
    "type": 505,
    "typeArguments": 5,
    "types": 6,
    "value": 6,
  },
  "DocTypeNameID": Object {
    "id": 90,
    "name": 114,
    "type": 114,
  },
}
`;

exports[`Types hsGraphD3 1`] = `
Object {
  "DocComment": Object {
    "returns": 9,
    "shortText": 192,
    "tags": 2,
    "text": 74,
  },
  "DocNode": Object {
    "children": 136,
    "comment": 182,
    "defaultValue": 96,
    "extendedBy": 15,
    "extendedTypes": 38,
    "flags": 1173,
    "getSignature": 24,
    "groups": 136,
    "id": 1173,
    "implementationOf": 24,
    "implementedBy": 1,
    "implementedTypes": 8,
    "indexSignature": 16,
    "inheritedFrom": 168,
    "kind": 1173,
    "kindString": 1171,
    "name": 1173,
    "originalName": 29,
    "overwrites": 56,
    "parameters": 16,
    "signatures": 215,
    "sources": 777,
    "type": 795,
  },
  "DocNodeFlags": Object {
    "isAbstract": 9,
    "isConst": 39,
    "isConstructorProperty": 6,
    "isExported": 458,
    "isOptional": 60,
    "isPrivate": 19,
    "isProtected": 73,
    "isPublic": 28,
    "isRest": 8,
    "isStatic": 14,
  },
  "DocReference": Object {
    "children": 231,
    "kind": 231,
    "title": 231,
  },
  "DocSignature": Object {
    "comment": 78,
    "flags": 215,
    "id": 215,
    "implementationOf": 15,
    "inheritedFrom": 34,
    "kind": 215,
    "kindString": 215,
    "name": 215,
    "overwrites": 43,
    "parameters": 182,
    "type": 215,
  },
  "DocSource": Object {
    "character": 778,
    "fileName": 778,
    "line": 778,
  },
  "DocTag": Object {
    "param": 3,
    "tag": 3,
    "text": 3,
  },
  "DocType": Object {
    "declaration": 33,
    "elementType": 101,
    "elements": 3,
    "id": 382,
    "name": 1024,
    "type": 1203,
    "typeArguments": 17,
    "types": 30,
    "value": 12,
  },
  "DocTypeNameID": Object {
    "id": 419,
    "name": 447,
    "type": 447,
  },
}
`;

exports[`Types hsLayout 1`] = `
Object {
  "DocComment": Object {
    "returns": 19,
    "shortText": 260,
    "tags": 3,
    "text": 103,
  },
  "DocNode": Object {
    "children": 165,
    "comment": 249,
    "defaultValue": 106,
    "extendedBy": 19,
    "extendedTypes": 46,
    "flags": 1371,
    "getSignature": 24,
    "groups": 165,
    "id": 1371,
    "implementationOf": 24,
    "implementedBy": 1,
    "implementedTypes": 8,
    "indexSignature": 16,
    "inheritedFrom": 200,
    "kind": 1371,
    "kindString": 1368,
    "name": 1371,
    "originalName": 38,
    "overwrites": 72,
    "parameters": 16,
    "signatures": 265,
    "sources": 915,
    "type": 913,
  },
  "DocNodeFlags": Object {
    "isAbstract": 14,
    "isConst": 43,
    "isConstructorProperty": 12,
    "isExported": 507,
    "isOptional": 65,
    "isPrivate": 27,
    "isProtected": 80,
    "isPublic": 49,
    "isRest": 8,
    "isStatic": 30,
  },
  "DocReference": Object {
    "children": 286,
    "kind": 286,
    "title": 286,
  },
  "DocSignature": Object {
    "comment": 105,
    "flags": 265,
    "id": 265,
    "implementationOf": 15,
    "inheritedFrom": 48,
    "kind": 265,
    "kindString": 265,
    "name": 265,
    "overwrites": 55,
    "parameters": 226,
    "type": 265,
  },
  "DocSource": Object {
    "character": 916,
    "fileName": 916,
    "line": 916,
  },
  "DocTag": Object {
    "param": 3,
    "tag": 4,
    "text": 4,
  },
  "DocType": Object {
    "declaration": 43,
    "elementType": 126,
    "elements": 3,
    "id": 443,
    "name": 1203,
    "type": 1425,
    "typeArguments": 30,
    "types": 38,
    "value": 12,
  },
  "DocTypeNameID": Object {
    "id": 509,
    "name": 537,
    "type": 537,
  },
}
`;

exports[`Types hsWidget 1`] = `
Object {
  "DocComment": Object {
    "returns": 20,
    "shortText": 329,
    "tags": 5,
    "text": 158,
  },
  "DocNode": Object {
    "children": 226,
    "comment": 328,
    "defaultValue": 144,
    "extendedBy": 22,
    "extendedTypes": 51,
    "flags": 1681,
    "getSignature": 24,
    "groups": 226,
    "id": 1681,
    "implementationOf": 24,
    "implementedBy": 1,
    "implementedTypes": 8,
    "indexSignature": 16,
    "inheritedFrom": 226,
    "kind": 1681,
    "kindString": 1677,
    "name": 1681,
    "originalName": 53,
    "overwrites": 86,
    "parameters": 16,
    "signatures": 348,
    "sources": 1113,
    "type": 1108,
  },
  "DocNodeFlags": Object {
    "isAbstract": 16,
    "isConst": 45,
    "isConstructorProperty": 12,
    "isExported": 678,
    "isOptional": 77,
    "isPrivate": 28,
    "isProtected": 92,
    "isPublic": 50,
    "isRest": 8,
    "isStatic": 58,
  },
  "DocReference": Object {
    "children": 357,
    "kind": 357,
    "title": 357,
  },
  "DocSignature": Object {
    "comment": 132,
    "flags": 348,
    "id": 348,
    "implementationOf": 15,
    "inheritedFrom": 74,
    "kind": 348,
    "kindString": 348,
    "name": 348,
    "overwrites": 69,
    "parameters": 309,
    "type": 348,
  },
  "DocSource": Object {
    "character": 1114,
    "fileName": 1114,
    "line": 1114,
  },
  "DocTag": Object {
    "param": 5,
    "tag": 7,
    "text": 7,
  },
  "DocType": Object {
    "declaration": 54,
    "elementType": 139,
    "elements": 3,
    "id": 533,
    "name": 1476,
    "type": 1727,
    "typeArguments": 31,
    "types": 43,
    "value": 12,
  },
  "DocTypeNameID": Object {
    "id": 599,
    "name": 627,
    "type": 627,
  },
}
`;
